server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/remote/delivery/MailDelivrer.java

```diff
@@ -118,14 +123,23 @@ private Domain retrieveTargetHostname(Mail mail) {
         return rcpt.getDomain();
     }
 
-    private ExecutionResult doDeliver(Mail mail, InternetAddress[] addr, Iterator<HostAddress> targetServers) throws MessagingException {
+    private ExecutionResult doDeliver(Mail mail, ImmutableSet<InternetAddress> addr, Iterator<HostAddress> targetServers) throws MessagingException {
         MessagingException lastError = null;
 
+        ImmutableSet<InternetAddress> targetAddresses = ImmutableSet.copyOf(addr);
+        ImmutableSet.Builder<InternetAddress> deliverySuccess = ImmutableSet.builder();
```

Comment:
I don't see the value of using an immutable collection for that given that we actually mutate it and we don't leak it outside this method

Reply:
I'm pretty sure the whole algorithm would be simpler that way:
```
HashSet<InternetAddress> targetAddresses = new HashSet(addr);
 while (targetServers.hasNext()) {
  return mailDelivrerToHost.tryDeliveryToHost(mail, addr, targetServers.next());
} catch (SendFailedException sfe) {
...
targetAddresses.removeAll(sfe.getValidSentAddresses());
...
}
```
WDYT?